internals/cli/masker/matcher.go

```diff
@@ -0,0 +1,135 @@
+package masker
+
+import (
+	"bytes"
+	"crypto/subtle"
+)
+
+// matches represents a set of sequence matches. The key is the index at which the match is found and the value is the
+// length of the match. The index corresponds to the index of the byte in the BufferedIndex of the stream.
+type matches map[int64]int
+
+// add a new match to the map if it does not yet exist or the existing match has a shorter length.
+func (m matches) add(index int64, length int) matches {
+	existing, exists := m[index]
+	if !exists || existing < length {
+		m[index] = length
+	}
+	return m
+}
+
+// matcher combines multiple sequenceMatchers to check for matches of secrets against any of them.
+type matcher struct {
+	detectors    []*sequenceDetector
+	currentIndex int64
+}
+
+// newMatcher returns a new matcher that contains a sequenceDetector for all given sequences.
+func newMatcher(sequences [][]byte) *matcher {
+	res := &matcher{
+		detectors: make([]*sequenceDetector, 0, len(sequences)),
+	}
+
+	for _, sequence := range sequences {
+		res.detectors = append(res.detectors, &sequenceDetector{
+			sequence: sequence,
+			offset:   0,
+		})
+		// Add detectors for any repetitions in the sequence.
+		// For example, if the sequence is "aaab", we also require a detector for "aaaab" and "aaaaab".
+		for length, count := range sequenceRepetitions(sequence) {
+			for i := 1; i <= count; i++ {
+				prefixedSequence := make([]byte, len(sequence)+length*i)
+				copy(prefixedSequence, sequence[:length*i])
+				copy(prefixedSequence[length*i:], sequence)
+				res.detectors = append(res.detectors, &sequenceDetector{
+					sequence: prefixedSequence,
+					offset:   length * i,
+				})
+			}
+		}
+	}
+	return res
+}
+
+// write takes in a slice of bytes and returns all matches found by any of its detectors.
+func (m *matcher) write(in []byte) matches {
+	res := matches{}
+	for i, b := range in {
+		for _, detector := range m.detectors {
+			match := detector.writeByte(b)
+			if match {
+				res = res.add(m.currentIndex+int64(i-detector.length()+1), detector.length())
+			}
+		}
+	}
+	m.currentIndex += int64(len(in))
+	return res
+}
+
+// sequenceDetector detects if a sequence is present in the bytes it receives.
+//
+type sequenceDetector struct {
+	sequence []byte
+	offset   int
+	index    int
+}
+
+// length returns the length of the sequence corrected for its offset.
+func (d *sequenceDetector) length() int {
+	return len(d.sequence) - d.offset
+}
+
+func (d *sequenceDetector) writeByte(in byte) bool {
+	// Implementation of the following code that limits data-dependency as much as possible.
+	//
+	// 	if d.sequence[d.index] == in {
+	// 		d.index++
+	//
+	//		if d.index == len(d.sequence) {
+	//			d.index = 0
+	//			return true
+	//		}
+	//		return false
+	//	} else if d.sequence[0] == in{
+	//		d.index = d.offset
+	//	} else {
+	//      d.index = 0
```

Comment:
Why does it start at zero and not at `d.offset` here?

Reply:
See example above.